modules/streams-core/src/main/scala/com.snowplowanalytics.snowplow/sources/SourceAndAck.scala

```diff
@@ -48,6 +49,22 @@ trait SourceAndAck[F[_]] {
    * healthy. If any event is "stuck" then latency is high and the probe should report unhealthy.
    */
   def isHealthy(maxAllowedProcessingLatency: FiniteDuration): F[SourceAndAck.HealthStatus]
+
+  /**
+   * Latency of the message that is currently being processed by the downstream application
+   *
+   * The returned value is `None` if this `SourceAndAck` is currently awaiting messages from
+   * upstream, e.g. it is doing a remote fetch.
+   *
+   * The returned value is `Some` if this `SourceAndAck` has emitted a message downstream to the
+   * application, and it is waiting for the downstream app to "pull" the next message from this
+   * `SourceAndAck`.
+   *
+   * This value should be used as the initial latency at the start of a statsd metrics reporting
+   * period. This ensures the app reports non-zero latency even when the app is stuck (e.g. cannot
+   * load events to destination).
+   */
+  def currentStreamLatency: F[Option[FiniteDuration]]
```

Comment:
As we put more and more latency stuff to common streams...maybe we could also do setting latency (when downstream pulls a batch) here? I mean, it looks the same pretty much everywhere, like:

```scala
  private def setLatency[F[_]: Sync](metrics: Metrics[F]): Pipe[F, TokenedEvents, TokenedEvents] =
    _.evalTap {
      _.earliestSourceTstamp match {
        case Some(t) =>
          for {
            now <- Sync[F].realTime
            latencyMillis = now.toMillis - t.toEpochMilli
            _ <- metrics.setLatencyMillis(latencyMillis)
          } yield ()
        case None =>
          Applicative[F].unit
      }
    }
``` 
is that right?

Can such code also live here? Then we would never have to expose `earliestSourceTstamp` to the app? Just an idea :)

Reply:
I took your idea and ran with it a bit.  It turned into quite a big change!  I'm happy with the results.

As well as the benefits you described (consolidate common implementation into one place) my newest commit makes it so the app can tell the difference between stuck stream (the source emits no metrics) vs healthy stream with no events to process (the source periodically emits zero latency).  The `latency_millis` metric should now be more reliable in how it triggers alerts when something is wrong.